compiler/passes/normalize.cpp

```diff
@@ -2833,18 +2833,76 @@ static void fixupExportedArrayFormals(FnSymbol* fn) {
         // TODO: handle call expression stuff.
       }
 
-      // Create a representation of the array argument that is accessible
-      // outside of Chapel (in the form of a pointer and a corresponding size)
-      formal->typeExpr->replace(new BlockStmt(new SymExpr(dtExternalArray->symbol)));
+      SymExpr* dom = toSymExpr(call->get(1));
+      if (dom != NULL && dom->symbol() == gNil) {
+        // The domain is nil.  Try to make a chpl_external_array with it.
+        // If that doesn't work, the user must be more explicit with their
+        // domain
+        formal->typeExpr->replace(new BlockStmt(new SymExpr(dtExternalArray->symbol)));
+      } else {
+        // Create a representation of the array argument that is accessible
+        // outside of Chapel, depending on the type of the array.  If the array
+        // is supported by our C/Python interop, it will be represented as a
+        // chpl_external_array.  Otherwise, it will be a chpl_opaque_array
+        VarSymbol* typeExprTmp = new VarSymbol("typeExpr");
+        typeExprTmp->addFlag(FLAG_MAYBE_TYPE);
+
+        CallExpr* newFormalType = new CallExpr("getExternalArrayType",
+                                               typeExprTmp);
+        formal->typeExpr->replace(new BlockStmt(new DefExpr(typeExprTmp)));
+        formal->typeExpr->insertAtTail(new CallExpr(PRIM_MOVE, typeExprTmp,
+                                                    call->copy()));
+        formal->typeExpr->insertAtTail(newFormalType);
+      }
+
+      VarSymbol* chplArr = new VarSymbol(astr(formal->name, "_arr"));
+      retCall->insertBefore(new DefExpr(chplArr));
 
       // Transform the outside representation into a Chapel array, and send that
       // in the call to the original function.
-      CallExpr* makeChplArray = new CallExpr("makeArrayFromExternArray",
+
+      // if (formal.type is dtExternalArray) then
```

Comment:
Is it possible for this to be one function call, like `chpl_helpMakeArrayFromExternOrOpaque(formal, eltExpr, oldTypeExpr)` ? Might make the compiler here a little simpler.

Reply:
The possibility of missing domains in the chpl__buildArrayRuntimeType call makes that tricky.  It means we won't be able to pass oldTypeExpr to the function in some cases, and detecting that will mean some of the complexity remains in the compiler.  I should probably move this into a helper function, though